### PR TITLE
YaoPlots style of Qobj generation

### PR DESCRIPTION
**Pull request by Sov-trotter**
So the previous approach was quite convoluted and hard to debug. This approach solves the issues via extending Mutiple Dispatch down to gates. 
A few todo's are 
- Cleaup the type defs (currently the Array and Tuple type defs are done to correctly dispatch to the controlled and pauli gates)
- Reduce repetitive code(better use Unions etc.)

### CODE REVIEW

**Comment by GiggleLiu** [comment]
I am a bit surprised to see both variational circuit and QFT circuit cam be compiled to IBMQ instantly. Well done!
@Roger-luo isn't this super cool!

## variational circuit

```julia
julia> using Yao, YaoExtensions, IBMQJulia

julia> c = variational_circuit(5);

julia> IBMQJulia.generate_instv2(c)
35-element Array{Any,1}:
 Any[Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [1]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [1])]
 Any[Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [2]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [2])]
 Any[Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [3]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [3])]
 Any[Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [4]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [4])]
 Any[Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [5]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [5])]
 Dict{String,Any}("name" => "cx","qubits" => [2, 1])
 Dict{String,Any}("name" => "cx","qubits" => [3, 2])
 Dict{String,Any}("name" => "cx","qubits" => [4, 3])
 Dict{String,Any}("name" => "cx","qubits" => [5, 4])
 Dict{String,Any}("name" => "cx","qubits" => [1, 5])
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [1]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [1]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [1])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [2]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [2]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [2])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [3]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [3]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [3])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [4]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [4]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [4])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [5]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [5]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [5])]
 Dict{String,Any}("name" => "cx","qubits" => [2, 1])
 Dict{String,Any}("name" => "cx","qubits" => [3, 2])
 Dict{String,Any}("name" => "cx","qubits" => [4, 3])
 Dict{String,Any}("name" => "cx","qubits" => [5, 4])
 Dict{String,Any}("name" => "cx","qubits" => [1, 5])
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [1]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [1]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [1])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [2]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [2]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [2])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [3]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [3]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [3])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [4]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [4]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [4])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [5]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [5]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [5])]
 Dict{String,Any}("name" => "cx","qubits" => [2, 1])
 Dict{String,Any}("name" => "cx","qubits" => [3, 2])
 Dict{String,Any}("name" => "cx","qubits" => [4, 3])
 Dict{String,Any}("name" => "cx","qubits" => [5, 4])
 Dict{String,Any}("name" => "cx","qubits" => [1, 5])
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [1]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [1])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [2]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [2])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [3]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [3])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [4]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [4])]
 Any[Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [5]), Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [5])]
```

## QFT Circuit

```julia
julia> IBMQJulia.generate_instv2(c)
15-element Array{Any,1}:
 Dict{String,Any}("name" => "h","qubits" => [1])
 Dict{String,Any}("name" => "cu1","params" => [1.5707963267948966],"qubits" => [1, 2])
 Dict{String,Any}("name" => "cu1","params" => [0.7853981633974483],"qubits" => [1, 3])
 Dict{String,Any}("name" => "cu1","params" => [0.39269908169872414],"qubits" => [1, 4])
 Dict{String,Any}("name" => "cu1","params" => [0.19634954084936207],"qubits" => [1, 5])
 Dict{String,Any}("name" => "h","qubits" => [2])
 Dict{String,Any}("name" => "cu1","params" => [1.5707963267948966],"qubits" => [2, 3])
 Dict{String,Any}("name" => "cu1","params" => [0.7853981633974483],"qubits" => [2, 4])
 Dict{String,Any}("name" => "cu1","params" => [0.39269908169872414],"qubits" => [2, 5])
 Dict{String,Any}("name" => "h","qubits" => [3])
 Dict{String,Any}("name" => "cu1","params" => [1.5707963267948966],"qubits" => [3, 4])
 Dict{String,Any}("name" => "cu1","params" => [0.7853981633974483],"qubits" => [3, 5])
 Dict{String,Any}("name" => "h","qubits" => [4])
 Dict{String,Any}("name" => "cu1","params" => [1.5707963267948966],"qubits" => [4, 5])
 Dict{String,Any}("name" => "h","qubits" => [5])
```

**Comment by GiggleLiu** [comment]
And the error message is more intuitive now.
```julia
julia> IBMQJulia.generate_instv2(c)
ERROR: MethodError: no method matching generate_instv2(::SqrtXGate, ::Array{Int64,1})
Closest candidates are:
  generate_instv2(::HGate, ::Any) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:45
  generate_instv2(::SWAPGate, ::Any) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:48
  generate_instv2(::XGate, ::Array) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:53
  ...
Stacktrace:
 [1] generate_instv2(::PutBlock{53,1,SqrtXGate}) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:38
 [2] generate_instv2(::ChainBlock{53}) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:30
 [3] _broadcast_getindex_evalf(::typeof(IBMQJulia.generate_instv2), ::ChainBlock{53}) at ./broadcast.jl:648
 [4] _broadcast_getindex(::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1},Tuple{Base.OneTo{Int64}},typeof(IBMQJulia.generate_instv2),Tuple{Base.Broadcast.Extruded{Array{AbstractBlock{53},1},Tuple{Bool},Tuple{Int64}}}}, ::Int64) at ./broadcast.jl:621
 [5] getindex(::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1},Tuple{Base.OneTo{Int64}},typeof(IBMQJulia.generate_instv2),Tuple{Base.Broadcast.Extruded{Array{AbstractBlock{53},1},Tuple{Bool},Tuple{Int64}}}}, ::Int64) at ./broadcast.jl:575
 [6] copy(::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1},Tuple{Base.OneTo{Int64}},typeof(IBMQJulia.generate_instv2),Tuple{Array{AbstractBlock{53},1}}}) at ./broadcast.jl:876
 [7] materialize(::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1},Nothing,typeof(IBMQJulia.generate_instv2),Tuple{Array{AbstractBlock{53},1}}}) at ./broadcast.jl:837
 [8] generate_instv2(::ChainBlock{53}) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:28
 [9] top-level scope at REPL[28]:1
```

So that user can specify these compilation process by themself.

File: src/qobj_v2.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    end
+    return inst
+end
+
+function generate_instv2(blk::PutBlock{N,M}) where {N,M}
+	locs = [blk.locs...]
+	generate_instv2(blk.content, locs)
+end
+
+function generate_instv2(blk::ControlBlock{N,GT,C}) where {N,GT,C}
```
Comment (GiggleLiu) [inline src/qobj_v2.jl]:
```julia
julia> c = control(5, -3, 2=>X)
nqubits: 5
control(¬3)
└─ (2,) X

julia> IBMQJulia.generate_instv2(c)
Dict{String,Any} with 2 entries:
  "name"   => "cx"
  "qubits" => [2, 3]

julia> c.ctrl_config
(0,)
```

Please check all variables in `ctrl_config` are `1`, and error on any `0`. The `0` represents the inverse control.

File: src/qobj_v2.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    for block in subblocks(qc_simpl)
+        # push!(inst, generate_instv2(block))
+        if block isa ChainBlock
+            push!(inst, generate_instv2.(block)...)
+        else
+            push!(inst, generate_instv2(block))
+        end
+    end
+    return inst
+end
```
Comment (GiggleLiu) [inline src/qobj_v2.jl]: Should be easy to support the following case by adding a `locs` argument
```julia
julia> c = put(5, 2=>chain(X, Y))
nqubits: 5
put on (2)
└─ chain
   ├─ X
   └─ Y


julia> IBMQJulia.generate_instv2(c)
ERROR: MethodError: no method matching generate_instv2(::ChainBlock{1}, ::Array{Int64,1})
Closest candidates are:
  generate_instv2(::HGate, ::Any) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:45
  generate_instv2(::SWAPGate, ::Any) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:48
  generate_instv2(::XGate, ::Array) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:53
  ...
Stacktrace:
 [1] generate_instv2(::PutBlock{5,1,ChainBlock{1}}) at /home/leo/.julia/dev/IBMQJulia/src/qobj_v2.jl:38
 [2] top-level scope at REPL[38]:1
```

**Comment by Sov-trotter** [comment]
The QFT circuit is pretty much correct for now, but the variational circuit needs to be rectified I guess. Thanks for this examples. At the end out instructions need to be `Array{Dict}` where as at some places it has done 
`Any[Dict{String,Any}("name" => "rx","params" => [0.0],"qubits" => [1]), Dict{String,Any}("name" => "rz","params" => [0.0],"qubits" => [1])]`. This is where the function sent back Array{DIct} for the nested chain block. I'll get it fixed

**Comment by GiggleLiu** [comment]
> the variational circuit needs to be rectified 

You are right, I didn't notice this issue. Maybe we need to add a method to validate the compiled circuit.
e.g.
```julia
julia> c = put(5, 2=>chain(X, Y))
nqubits: 5
put on (2)
└─ chain
   ├─ X
   └─ Y

julia> compiled_c = compile_to_ibmq(c)  # return a block of type `IBMQobj`
nqubits: 5
IBMQobj
├─ 2=>x
└─ 2=>y

julia> @test mat(compiled_c) ≈ mat(c)  # compare their matrices
pass
```

Please refer the `mat` and `apply!` functions of `ChainBlock`
https://github.com/QuantumBFS/YaoBlocks.jl/blob/6d3fc833d431df47a7430a1dd341dc9347ffca2e/src/composite/chain.jl#L77

and `PutBlock`
https://github.com/QuantumBFS/YaoBlocks.jl/blob/fd0d820d1f67c8c5413bf4b7492ed1550d6763ea/src/composite/put_block.jl#L76

Or equivalently, just transform the Qobj to a chain of put blocks, and then call `mat`.

**Comment by Sov-trotter** [comment]
> Maybe we need to add a method to validate the compiled circuit.

Yeah. There's actually a way around that. Where the generated json is compared with the standard json schema of IBMQ Qobj. 
BTW I like what you suggested. This way we can assure that everything that's going in is correct.
But that's something I wish to tackle later on first we can try to get the things right and support all the gates and stuff?
What do you think?

**Comment by GiggleLiu** [comment]
Agree that should be included in a seperate PR.

File: src/qobj.jl [hunk 3]
```diff
@@ -24,92 +26,60 @@ end
...
-            nm = "ry"
-        elseif super_blk isa XGate
-            nm = "rx"
-        end
-        inst = Dict("name"=>"$(nm)", "qubits"=>[blk.locs...], "params"=>[gate.theta])
-        push!(data, inst)
-    end
-    return Tuple(data)
+function generate_inst(blk::ControlBlock{N,GT,C}) where {N,GT,C}
+	generate_instv2(blk.content, blk.locs, blk.ctrl_locs)
```
Comment (GiggleLiu) [inline src/qobj.jl]: Is this a legacy code?

Comment (Sov-trotter) [reply src/qobj.jl]: No. It handles the dispatch for control block.

Comment (GiggleLiu) [reply src/qobj.jl]: I mean the `v2`.

**Comment by GiggleLiu** [comment]
Perfect!

**Pull request merged**